Graphics/GraphicsTools/src/BytecodeCache.cpp

```diff
@@ -0,0 +1,221 @@
+{
+public:
+    using TBase = ObjectBase<IBytecodeCache>;
+
+    struct BytecodeCacheHeader
+    {
+        Version CacheVersion  = {};
+        Uint64  CountElements = 0;
+    };
+
+    struct BytecodeCacheElementHeader
+    {
+        XXH128Hash Hash     = {};
+        Uint64     DataSize = 0;
+    };
+
+    static constexpr Uint64 InitialStreamSize = 4096;
+
+public:
+    BytecodeCacheImpl(IReferenceCounters*            pRefCounters,
+                      const BytecodeCacheCreateInfo& CreateInfo) :
+        TBase{pRefCounters},
+        m_DeviceType{CreateInfo.DeviceType}
+    {
+    }
+
+    IMPLEMENT_QUERY_INTERFACE_IN_PLACE(IID_BytecodeCache, TBase);
+
+    virtual void DILIGENT_CALL_TYPE Load(IDataBlob* pDataBlob) override final
+    {
+        VERIFY_EXPR(pDataBlob != nullptr);
+        InputMemoryStream Stream{static_cast<const uint8_t*>(pDataBlob->GetConstDataPtr()), pDataBlob->GetSize()};
+
+        BytecodeCacheHeader Header{};
+        Stream.Read(Header);
+
+        for (Uint64 ItemID = 0; ItemID < Header.CountElements; ItemID++)
+        {
+            BytecodeCacheElementHeader ElementHeader{};
+            Stream.Read(ElementHeader);
+
+            auto pBytecode = DataBlobImpl::Create(StaticCast<size_t>(ElementHeader.DataSize));
+            Stream.Read(pBytecode->GetDataPtr(), ElementHeader.DataSize);
+            m_HashMap.emplace(ElementHeader.Hash, pBytecode);
+        }
+    }
+
+    virtual void DILIGENT_CALL_TYPE GetBytecode(const ShaderCreateInfo& ShaderCI, IDataBlob** ppByteCode) override final
+    {
+        VERIFY_EXPR(ppByteCode != nullptr);
+        const auto Hash = ComputeHash(ShaderCI);
+        const auto Iter = m_HashMap.find(Hash);
+        if (Iter != m_HashMap.end())
+        {
+            auto pObject = Iter->second;
+            *ppByteCode  = pObject.Detach();
+        }
+    }
+
+    virtual void DILIGENT_CALL_TYPE AddBytecode(const ShaderCreateInfo& ShaderCI, IDataBlob* pByteCode) override final
+    {
+        VERIFY_EXPR(pByteCode != nullptr);
+        const auto Hash = ComputeHash(ShaderCI);
+        const auto Iter = m_HashMap.emplace(Hash, pByteCode);
+        if (!Iter.second)
+            Iter.first->second = pByteCode;
+    }
+
+    virtual void DILIGENT_CALL_TYPE RemoveBytecode(const ShaderCreateInfo& ShaderCI) override final
+    {
+        const auto Hash = ComputeHash(ShaderCI);
+        m_HashMap.erase(Hash);
+    }
+
+    virtual void DILIGENT_CALL_TYPE Store(IDataBlob** ppDataBlob) override final
+    {
+        VERIFY_EXPR(ppDataBlob != nullptr);
+        OutputMemoryStream Stream{InitialStreamSize};
+
+        BytecodeCacheHeader Header{};
+        Header.CountElements = m_HashMap.size();
+        Stream.Write(Header);
+
+        for (auto const& Pair : m_HashMap)
+        {
+            Stream.Write(BytecodeCacheElementHeader{Pair.first, Pair.second->GetSize()});
+            Stream.Write(Pair.second->GetConstDataPtr(), Pair.second->GetSize());
+        }
```

Comment:
I suggest using Serializer here. This way you will not need a new OutputStream class